mp1_assignment/mp1/MP1Node.cpp

```diff
@@ -215,9 +222,79 @@ void MP1Node::checkMessages() {
  * DESCRIPTION: Message handler for different message types
  */
 bool MP1Node::recvCallBack(void *env, char *data, int size ) {
-	/*
-	 * Your code goes here
-	 */
+    if(memberNode->bFailed) {
+        return false;
+    }
+	MessageHdr* msg = (MessageHdr *) data;
+    // Consume msg and construct response
+
+    if (msg->msgType == JOINREQ){
+        pushMember(msg);
+        
+        MessageHdr* replyMsg = new MessageHdr();
+        replyMsg->msgType = JOINREP;
+        replyMsg->members = memberNode->memberList;
+        replyMsg->addr = &memberNode->addr;
+
+        // reply with JOINREP message 
+        emulNet->ENsend(&memberNode->addr, msg->addr, (char *)replyMsg, sizeof(MessageHdr));
+    }
+    else if(msg->msgType == JOINREP) {
+        pushMember(msg);
+        memberNode->inGroup = true;
+    }
+    else if(msg->msgType == PING) {
+        handlePing(msg);
+    }
+    return true;
+}
+
+void MP1Node::handlePing(MessageHdr* msg) {
+    for(const auto& updateMember: msg->members) {
+        MemberListEntry* updateEntry = check_member_list(updateMember.id, updateMember.port); 
+        if(updateMember.timestamp + TREMOVE < par->getcurrtime()) {
```

Comment:
this was the part I missed that cause accuracy check failure. if the update we get from ping msg is old, we shouldn't add it and propergate it.